### PR TITLE
fix(loop): scope retrospective recency to merged PRs and accept reconcile envelope

### LINKED ISSUE #2027: Fix retrospective PR recency and reconciliation-envelope bootstrap dead end
## Context

The startup resolver enforces the repository-wide behavioral-retrospective checkpoint before a qualifying dev-loop start or resume. A completed or explicitly skipped checkpoint is scoped to the PR merge recorded in its cycle identity.

Two coupled bugs currently make the documented startup path unusable:

1. `resolveHasNewerMergeSinceCheckpoint` treats any commit after the checkpoint SHA on `origin/<baseBranch>` as a newer retrospective cycle. In the observed case, the valid completed checkpoint for PR #2019 at `5868c09a8ebb360975449809a31b6e0b7fa49c96` was followed only by direct release commit `0649cd1e`; startup incorrectly returned `needs_reconcile` for issue #2027.
2. The resulting legitimate fail-closed startup payload cannot be consumed by `loop build-envelope`: `routeKind`/`bundleKind` are `needs_reconcile` and `selectedStrategy` is null, but envelope validation rejects it with `resolverOutput.selectedStrategy is required and must be a non-empty string` instead of producing the actionable reconciliation envelope required by the public workflow.

Together these form a bootstrap dead end: `dev-loop` cannot start the issue that fixes its false checkpoint block, and its mandatory handoff-envelope step cannot represent the stop correctly.

This issue owns PR-only retrospective-recency classification and reconciliation-envelope compatibility. It does NOT own broader retrospective qualification, provenance, checkpoint lifecycle, or gate policy (#1967).

The same live fix cycle exposed a PR-authoring feedback loop: validation evidence in PR #2028 included aggregate assertion, asset, skip, and test counts. Those incidental totals changed as coverage improved, mutating the PR-body/spec digest and forcing otherwise avoidable re-review fan-out. This issue therefore also owns the narrow durable PR-description rule needed to keep validation evidence stable across implementation revisions.

## Decided approach

Retrospective recency means “a newer PR was merged into the configured base branch,” not “the base branch contains any newer commit” and not “Git history contains a two-parent merge commit.” Because this repository squash-merges, classify commits after the checkpoint using authoritative GitHub PR association and merge/base-ref facts. Ignore commits with no qualifying merged-PR association.

Keep the existing best-effort base-branch refresh and ancestry boundary. A newer commit associated with a PR merged into the configured base branch makes the checkpoint stale. Direct commits, unmerged PRs, and PRs merged into another base do not. If recorded ancestry or required association facts cannot be verified, preserve the fail-closed `missing` outcome.

Separately, make `loop build-envelope` accept the router's intentional no-strategy terminal/reconciliation states. A `needs_reconcile` resolver output with `selectedStrategy: null` must yield an envelope carrying its required reads, next action, stop rules, and acceptance contract. This is not a gate exemption or permission to route implementation.

For tracker-backed PR descriptions, record validation evidence as the command or named check plus a stable pass/fail outcome. Keep volatile aggregate test/assertion/asset counts, skip counts, durations, timestamps, and similar incidental run totals in head-stamped validation and gate artifacts rather than the PR body, unless an acceptance criterion explicitly makes an exact quantity behaviorally significant. Pin this rule in the canonical PR-description contract, pull-request template, generated/mirrored guidance, and deterministic contract coverage without creating an overbroad numeric validator.

## AC / DoD matrix

| Acceptance-criterion outcome | Required completion evidence |
| --- | --- |
| A direct commit on the configured base branch after a valid `complete` checkpoint does not make it stale. | Startup-level regression models PR #2019 followed only by the direct release commit and proves routing proceeds past retrospective reconciliation. |
| A direct commit after a valid `skipped` checkpoint likewise remains satisfied. | State-resolution or startup coverage proves all newer commits without a qualifying merged-PR association are ignored. |
| A squash-merged PR after the checkpoint makes it stale despite having one parent. | Coverage supplies authoritative merged-PR association for a newer commit on the configured base and expects `missing` plus `needs_reconcile`. |
| Mixed history is classified correctly. | Tests cover direct-only and direct-plus-qualifying-PR histories without relying on `git log --merges`. |
| Only PRs merged into the configured base branch qualify. | Tests distinguish the configured base from open, closed-unmerged, and differently based PR associations. |
| Association or ancestry uncertainty fails closed. | Tests cover an unresolvable checkpoint SHA and GitHub association/API failure, retaining `missing`/`needs_reconcile`. |
| A legitimate reconciliation result is envelope-compatible. | `loop build-envelope` accepts `bundleKind`/`routeKind: needs_reconcile` with `selectedStrategy: null` and emits an actionable reconciliation envelope; routed strategies still require a non-empty strategy. |
| The documented end-to-end startup path is unblocked. | Package/CLI-level regression runs startup state through `loop build-envelope`, proving a direct-only history reaches its real route and a genuine reconciliation result yields a valid stop envelope; pure helper-only evidence is insufficient. |
| Existing checkpoint and routing semantics remain unchanged outside this fix. | Coverage retains opt-in behavior, absent file → `none`, malformed records → `missing`, fresh-context provenance for `complete`, and no gate bypass. |
| Operator-facing contracts precisely describe PR-only recency and nullable strategy for terminal reconciliation. | Canonical contract prose, inline guidance, and generated mirrors are updated consistently. |
| The fix remains harness-agnostic. | Focused core/CLI tests and `bun run verify` pass under pinned Bun 1.4.1; cross-harness regression coverage prevents Pi/Claude/Codex divergence. |
| Tracker-backed PR validation evidence remains stable across ordinary assertion and asset-count drift. | Canonical PR-description prose and the pull-request template require command/check plus stable pass/fail outcomes; volatile counts/durations/timestamps stay in head-stamped validation/gate artifacts unless an acceptance criterion makes an exact quantity behaviorally significant; deterministic contract coverage pins prose/template parity. |

## Non-goals

- Changing which dev-loop strategies or completions qualify for retrospectives.
- Treating only two-parent Git merge commits as PR merges.
- Requiring retrospectives for direct commits, releases, tags, or other non-PR base updates.
- Changing fresh-context provenance, checkpoint storage, explicit-skip semantics, or the fail-closed gate.
- Automatically rewriting, clearing, fabricating, or bypassing an operator checkpoint.
- Allowing a null strategy for successful routed work.
- Changing release automation or prohibiting direct base-branch commits.
- Rejecting all numeric validation evidence or prohibiting exact quantities that are behaviorally significant to an explicit acceptance criterion.

### PR DESCRIPTION
**Pull request by mfittko**
## Summary

Fixes the bootstrap dead end where `dev-loop` could not start the very issue that repairs its false retrospective-checkpoint block, and its mandatory handoff-envelope step could not represent the stop.

## In scope

- Scope retrospective staleness to PRs merged into the configured base branch, not any newer commit. Direct commits, release commits, tags, open/closed-unmerged PRs, and PRs merged into another base no longer make a checkpoint stale. Classification uses the authoritative GitHub `Commit.associatedPullRequests` GraphQL connection; unverifiable ancestry or association fails closed.
- Make `loop build-envelope` accept the router's intentional terminal reconciliation tuple (`routeKind: needs_reconcile`, `selectedGate: fail_closed_reconcile`, `selectedStrategy: null`) and emit an actionable reconciliation envelope. Routed strategies still require a non-empty strategy.
- Add the durable `OPS-PR-VALIDATION-STABLE-EVIDENCE` PR-description rule: validation evidence records command/check plus a stable pass/fail outcome; volatile aggregate counts, durations, and timestamps stay in head-stamped validation/gate artifacts.

## Acceptance criteria

- [x] A direct commit on the configured base after a valid `complete` checkpoint does not make it stale.
- [x] A direct commit after a valid `skipped` checkpoint likewise remains satisfied.
- [x] A squash-merged PR into the configured base after the checkpoint makes it stale despite one parent.
- [x] Mixed history is classified correctly without relying on `git log --merges`.
- [x] Only PRs merged into the configured base branch qualify (open, closed-unmerged, differently-based PRs do not).
- [x] Association or ancestry uncertainty fails closed (`missing`/`needs_reconcile`).
- [x] A legitimate reconciliation result builds and validates an actionable terminal envelope; routed strategies still require a non-empty strategy.
- [x] The documented startup-to-envelope path is covered end to end at the package/CLI level.
- [x] Existing checkpoint/routing semantics are unchanged outside this fix (opt-in, absent→none, malformed→missing, fresh-context provenance for complete, no gate bypass).
- [x] Operator-facing contracts describe PR-only recency and nullable strategy for terminal reconciliation, with generated mirrors synchronized.
- [x] Tracker-backed PR validation evidence remains stable across incidental count drift (contract prose + template + deterministic coverage).

## Definition of done

- [x] Focused core and CLI regression tests pass.
- [x] Canonical contract prose and generated `.claude/` mirrors are synchronized.
- [x] `bun run verify` passes under pinned Bun.
- [x] User-facing behavior recorded under `CHANGELOG.md` Unreleased.

## Non-goals

- Changing which dev-loop strategies/completions qualify for retrospectives.
- Treating only two-parent Git merges as PR merges, or treating direct commits/releases/tags as PR merges.
- Changing fresh-context provenance, checkpoint storage, explicit-skip semantics, or the fail-closed gate.
- Automatically rewriting, clearing, or bypassing an operator checkpoint.
- Allowing a null strategy for successful routed work.
- Rejecting numeric evidence that an acceptance criterion makes behaviorally significant.
- Reverting the stale `workflow.baseBranch` config override (tracked separately as #2094).

## Validation

- `bun test packages/core/test/handoff-envelope.test.mjs test/loop/build-handoff-envelope.test.mjs test/loop/resolve-dev-loop-startup.test.mjs test/contracts/copilot-review-doc-contracts.test.mjs` — pass
- `node scripts/claude/generate-claude-assets.mjs --check` — pass
- `bun run verify` — pass

Closes #2027

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/loop/resolve-dev-loop-startup.mjs [hunk 1]
```diff
@@ -346,7 +447,48 @@ export function resolveHasNewerMergeSinceCheckpoint({ mergeCommit, baseBranch, c
...
+      }
+      const mergeCommitSha = association.merge_commit_sha;
+      if (associationState === "MERGED") {
+        if (
+          !mergedAt
+          || typeof mergeCommitSha !== "string"
+          || mergeCommitSha.trim() !== mergeCommitSha
+          || mergeCommitSha.length === 0
+        ) return true;
+        if (associationBase === baseBranch && mergeCommitSha === commitSha) return true;
```
Comment (mfittko) [inline scripts/loop/resolve-dev-loop-startup.mjs:485]: **low** (`coverage`): The merge-commit identity requirement mergeCommitSha === commitSha is not isolated by any test. Every MERGED-into-base fixture (squash-stale, mixed-history) sets merge_commit_sha equal to the candidate commit, and the only differing-sha fixture (46) also uses a non-base ref (release), so an AND->base-only mutation of this line would pass the whole suite. The branch where a PR merged INTO the configured base but with a different merge commit does NOT qualify is untested, so a false-positive-staleness regression here is uncaught.

File: scripts/loop/resolve-dev-loop-startup.mjs [hunk 2]
```diff
@@ -310,22 +310,111 @@ function ghJson(args, cwd, env = process.env) {
...
+    if (!connection.pageInfo.hasNextPage) return associations;
+    after = connection.pageInfo.endCursor;
+    if (typeof after !== "string" || after.length === 0) {
+      throw new Error("GraphQL commit-to-PR association pagination cursor is missing");
+    }
+  }
+  throw new Error(`GraphQL commit-to-PR association exceeded ${COMMIT_ASSOCIATION_MAX_PAGES} pages`);
+}
+
+function isStrictGitHubRfc3339Timestamp(value) {
```
Comment (mfittko) [inline scripts/loop/resolve-dev-loop-startup.mjs:369]: **nit** (`coverage`): isStrictGitHubRfc3339Timestamp carries calendar arithmetic (leap-year, days-in-month, hour/minute/offset bounds) but tests exercise only 'not-a-date' and a numeric value. The calendar-invalid branches (e.g. month 13, Feb 30, offset 24:00) are untested. These only tighten rejection and fail closed, so the risk is low; the gap is really that the extra validation logic is more than the observed inputs need.

File: scripts/loop/resolve-dev-loop-startup.mjs [hunk 3]
```diff
@@ -310,22 +310,111 @@ function ghJson(args, cwd, env = process.env) {
...
+  }
+}`;
+function resolveCommitPullRequestsFromGitHub({ repo, commitSha, cwd, env }) {
+  if (typeof repo !== "string" || repo.trim().length === 0) {
+    throw new Error("repository identity is required for commit-to-PR association");
+  }
+  // GraphQL is the sole merge authority: unlike the REST commit-PR endpoint it
+  // reports associations against any base branch. Bound pagination and fail
+  // closed when the complete connection cannot be verified.
+  const [owner, name] = repo.split("/");
```
Comment (mfittko) [inline scripts/loop/resolve-dev-loop-startup.mjs:333]: **nit** (`input-validation`): resolveCommitPullRequestsFromGitHub parses the repo slug with ad-hoc const &#91;owner, name] = repo.split("/") instead of the shared parseRepoSlugParts already exported from the repo-slug module this file imports. No live hazard: repo is currentRepo from detectRepoSlug, whose regex guarantees a single-slash owner/name shape, so the split is always well-formed here.

**Review by mfittko — COMMENTED** [review]
### Gate review: `draft_gate`


**Reviewed head SHA:** `ad4f5ea03b00e6773965faf1457fe6a5c61491e7`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 11 angles reviewed; 3 findings (see per-angle breakdown below).

**Inline findings:** 3 (🟡 1 low · ⚪ 2 nit) — on the diff, see Files changed. Angles: coverage, input-validation.

**Clean (9):** ci-guard, config-drift, contract-surface, correctness, determinism, gate-evidence, link-check, pr-description, scope

**Next action:** Move to ready-for-review; run pre_approval_gate

File: scripts/loop/resolve-dev-loop-startup.mjs (hunk 1 above)
Comment (mfittko) [reply scripts/loop/resolve-dev-loop-startup.mjs:485]: Resolved at gate close (round 1, fingerprint ad09a7ce2813f543, severity low, angle coverage): Examined on merits: "The merge-commit identity requirement mergeCommitSha === commitSha is not isolated by any test. Every MERGED-into-base fixture (squash-stale, mixed-history) sets merge_commit_sha equal to the candidate commit, and the only differing-sha fixture (46) also uses a non-base ref (release), so an AND->base-only mutation of this line would pass the whole suite. The branch where a PR merged INTO the configured base but with a different merge commit does NOT qualify is untested, so a false-positive-staleness regression here is uncaught." was reviewed against current scope and acceptance criteria; fixer triage found no same-commit fix warranted and the finding does not clear the operator-visibility filing bar. this low finding carries no operator-visibility signal, so it is resolved with rationale at gate close instead of filed to a tracked follow-up issue (net-reduction disposition policy).

File: scripts/loop/resolve-dev-loop-startup.mjs (hunk 2 above)
Comment (mfittko) [reply scripts/loop/resolve-dev-loop-startup.mjs:369]: Resolved at gate close (round 1, fingerprint bf04486ef20e3eba, severity nit, angle coverage): Examined on merits: "isStrictGitHubRfc3339Timestamp carries calendar arithmetic (leap-year, days-in-month, hour/minute/offset bounds) but tests exercise only 'not-a-date' and a numeric value. The calendar-invalid branches (e.g. month 13, Feb 30, offset 24:00) are untested. These only tighten rejection and fail closed, so the risk is low; the gap is really that the extra validation logic is more than the observed inputs need." was reviewed against current scope and acceptance criteria; it is cosmetic and does not warrant a fixer cycle or tracked follow-up. nit findings are resolved with rationale at gate close, with no fixer cycle and no tracked follow-up issue (net-reduction disposition policy).

File: scripts/loop/resolve-dev-loop-startup.mjs (hunk 3 above)
Comment (mfittko) [reply scripts/loop/resolve-dev-loop-startup.mjs:333]: Resolved at gate close (round 1, fingerprint 4dbd7735fdbe69ca, severity nit, angle input-validation): Examined on merits: "resolveCommitPullRequestsFromGitHub parses the repo slug with ad-hoc const &#91;owner, name] = repo.split('/') instead of the shared parseRepoSlugParts already exported from the repo-slug module this file imports. No live hazard: repo is currentRepo from detectRepoSlug, whose regex guarantees a single-slash owner/name shape, so the split is always well-formed here." was reviewed against current scope and acceptance criteria; it is cosmetic and does not warrant a fixer cycle or tracked follow-up. nit findings are resolved with rationale at gate close, with no fixer cycle and no tracked follow-up issue (net-reduction disposition policy).

**Review by mfittko — COMMENTED** [review]
### Gate review: `draft_gate`


**Reviewed head SHA:** `93d964d5cda8b13fa94f8704f00ac7d6e928c382`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 11 angles reviewed; no findings (see per-angle breakdown below).

**Clean (11):** ci-guard, config-drift, contract-surface, correctness, coverage, determinism, gate-evidence, input-validation, link-check, pr-description, scope

**Next action:** Move to ready-for-review; run pre_approval_gate

File: test/loop/build-handoff-envelope.test.mjs [hunk 4]
```diff
@@ -177,6 +177,38 @@ test("build-handoff-envelope builds envelope from resolver output", async () =>
...
+      nextAction: "Complete the required retrospective before routing.",
+      bundle: {
+        ...makeResolverOutput().bundle,
+        bundleKind: "needs_reconcile",
+        routeKind: "needs_reconcile",
+        selectedGate: "fail_closed_reconcile",
+        selectedStrategy: null,
+        loopState: "unknown",
+        nextAction: "Complete the required retrospective before routing.",
+      },
```
[the remark this answers is not in the document]
Comment (mfittko) [reply test/loop/build-handoff-envelope.test.mjs:195]: Fixed in 95275529: both nextAction fields (and the assertion) in this build-envelope fixture now use the canonical retrospective directive ("Complete or explicitly skip the required post-run behavioral retrospective before starting or resuming the next dev-loop run."), which matches validateHandoffEnvelope's fail-closed prefix allowlist. The fixture now exercises a realistic reconcile nextAction and would pass validation if the test is later extended to validate the envelope.

File: scripts/loop/resolve-dev-loop-startup.mjs (hunk 2 above)
Comment (mfittko) [inline scripts/loop/resolve-dev-loop-startup.mjs:369]: **low** (`yagni`): isStrictGitHubRfc3339Timestamp does full calendar validation (leap-year rule, per-month day counts, offset bounds), but mergedAt is only ever used as a truthiness gate for a MERGED association and is never compared as a date. The extra calendar math changes no outcome the AC or tests exercise; a non-empty-string check plus Date.parse would fail closed identically.

**Review by mfittko — COMMENTED** [review]
### Gate review: `pre_approval_gate`


**Reviewed head SHA:** `95275529bc48dc7923c48bafd481f6a60ab9d0b5`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 7 angles reviewed; 1 finding (see per-angle breakdown below).

**Inline findings:** 1 (🟡 1 low) — on the diff, see Files changed. Angles: yagni.

**Clean (6):** acceptance-criteria, contradiction-lens, docs, dry, pr-checklist, renderer-security

**Gate evidence note:** Copilot review rounds exhausted (2/2); current head has zero unresolved threads and green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

**Next action:** Merge-ready: draft_gate + pre_approval_gate clean at current head, CI green, zero unresolved threads, Copilot converged (round_cap_clean_fallback). Awaiting human merge per operator merge policy.

File: scripts/loop/resolve-dev-loop-startup.mjs (hunk 2 above)
Comment (mfittko) [reply scripts/loop/resolve-dev-loop-startup.mjs:369]: Deferred (disposition: deferred, non-blocking low). isStrictGitHubRfc3339Timestamp performs full calendar validation, but mergedAt is used only as a truthiness gate for a MERGED association, never compared as a date — so the extra strictness changes no behavior and fails closed identically to a non-empty-string + Date.parse check. Not a correctness/AC defect; resolved with rationale rather than fixed to avoid a head-bump re-gate. The pre_approval_gate blocks on high only; this low does not block. Recorded in the pre_approval findings-log.

File: packages/core/src/loop/handoff-envelope.mjs [hunk 5]
```diff
@@ -832,6 +887,58 @@ export function validateHandoffEnvelope(envelope) {
...
+        || (envelope.stopRules.length === 2 && envelope.stopRules[1] === "merge"));
+    // Routing invariants only: acceptance prose remains the resolver's
+    // business, while nextAction must begin with one of the known fail-closed
+    // directives rather than arbitrary non-empty text.
+    const reconciliationCriterion = envelope.acceptance?.criteria;
+    const reconciliationAcceptanceIsCanonical = Array.isArray(reconciliationCriterion)
+      && reconciliationCriterion.some((criterion) => criterion?.id === "reconcile" && criterion?.severity === "required");
+    const normalizedNextAction = typeof envelope.nextAction === "string" ? envelope.nextAction.trim() : "";
+    const reconciliationNextActionIsActionable = [
+      "Reconcile ",
```
Comment (mfittko) [inline packages/core/src/loop/handoff-envelope.mjs:910]: **low** (`dry`): The validator's reconciliation nextAction allowlist duplicates directive-string prefixes ("Stop and reconcile ", "Complete or explicitly skip ", "Local implementation requires worktree isolation") that are authored independently in four producer modules (public-dev-loop-routing.mjs, conductor-routing.mjs, retrospective-checkpoint.mjs, resolve-dev-loop-startup.mjs) with no shared constant, so the vocabulary is restated rather than referenced.

File: packages/core/src/loop/handoff-envelope.mjs [hunk 6]
```diff
@@ -832,6 +887,58 @@ export function validateHandoffEnvelope(envelope) {
...
+      && (envelope.stopRules.length === 1
+        || (envelope.stopRules.length === 2 && envelope.stopRules[1] === "merge"));
+    // Routing invariants only: acceptance prose remains the resolver's
+    // business, while nextAction must begin with one of the known fail-closed
+    // directives rather than arbitrary non-empty text.
+    const reconciliationCriterion = envelope.acceptance?.criteria;
+    const reconciliationAcceptanceIsCanonical = Array.isArray(reconciliationCriterion)
+      && reconciliationCriterion.some((criterion) => criterion?.id === "reconcile" && criterion?.severity === "required");
+    const normalizedNextAction = typeof envelope.nextAction === "string" ? envelope.nextAction.trim() : "";
+    const reconciliationNextActionIsActionable = [
```
Comment (mfittko) [inline packages/core/src/loop/handoff-envelope.mjs:909]: **low** (`yagni`): validateHandoffEnvelope enforces a hardcoded 4-prefix allowlist for the reconciliation nextAction ('Reconcile ', 'Stop and reconcile ', 'Complete or explicitly skip ', 'Local implementation requires worktree isolation'). No AC requires phrasing-level validation of nextAction; workflow-handoff-contract.md itself states the resolver is the authority for the actual nextAction and the builder copies it unchanged. Encoding the resolver's exact phrasings into the validator couples the two components: any new or reworded reconcile directive silently fails validation until this list is edited, and the check only defends against a hand-tampered envelope, a speculative threat inside a single trusted build->validate pipeline.

**Review by mfittko — COMMENTED** [review]
### Gate review: `pre_approval_gate`


**Reviewed head SHA:** `e52515f83df1a85453e87b59838b5bea28838822`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 7 angles reviewed; 2 findings (see per-angle breakdown below).

**Inline findings:** 2 (🟡 2 low) — on the diff, see Files changed. Angles: dry, yagni.

**Clean (5):** acceptance-criteria, contradiction-lens, docs, pr-checklist, renderer-security

**Gate evidence note:** Copilot review rounds exhausted (2/2); current head has zero unresolved threads and green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

**Next action:** Merge-ready: draft_gate + pre_approval_gate clean at current head, CI green, zero unresolved threads, Copilot converged (round_cap_clean_fallback). Awaiting human merge per operator merge policy.

File: packages/core/src/loop/handoff-envelope.mjs (hunk 5 above)
Comment (mfittko) [reply packages/core/src/loop/handoff-envelope.mjs:910]: Resolved at gate close (round 2, fingerprint ab44c4537de28103, severity low, angle dry): Examined on merits: "The validator's reconciliation nextAction allowlist duplicates directive-string prefixes ('Stop and reconcile ', 'Complete or explicitly skip ', 'Local implementation requires worktree isolation') that are authored independently in four producer modules (public-dev-loop-routing.mjs, conductor-routing.mjs, retrospective-checkpoint.mjs, resolve-dev-loop-startup.mjs) with no shared constant, so the vocabulary is restated rather than referenced." was reviewed against current scope and acceptance criteria; fixer triage found no same-commit fix warranted and the finding does not clear the operator-visibility filing bar. this low finding carries no operator-visibility signal, so it is resolved with rationale at gate close instead of filed to a tracked follow-up issue (net-reduction disposition policy).

File: packages/core/src/loop/handoff-envelope.mjs (hunk 6 above)
Comment (mfittko) [reply packages/core/src/loop/handoff-envelope.mjs:909]: Resolved at gate close (round 2, fingerprint a505fc3fe70ac013, severity low, angle yagni): Examined on merits: "validateHandoffEnvelope enforces a hardcoded 4-prefix allowlist for the reconciliation nextAction ('Reconcile ', 'Stop and reconcile ', 'Complete or explicitly skip ', 'Local implementation requires worktree isolation'). No AC requires phrasing-level validation of nextAction; workflow-handoff-contract.md itself states the resolver is the authority for the actual nextAction and the builder copies it unchanged. Encoding the resolver's exact phrasings into the validator couples the two components: any new or reworded reconcile directive silently fails validation until this list is edited, and the check only defends against a hand-tampered envelope, a speculative threat inside a single trusted build->validate pipeline." was reviewed against current scope and acceptance criteria; fixer triage found no same-commit fix warranted and the finding does not clear the operator-visibility filing bar. this low finding carries no operator-visibility signal, so it is resolved with rationale at gate close instead of filed to a tracked follow-up issue (net-reduction disposition policy).
